core/src/main/java/org/apache/iceberg/CachingCatalog.java

```diff
@@ -20,33 +20,143 @@
+   * by the catalog and not the cache itself.
+   * <p>
+   * Returns the age of the cache entry corresponding to the identifier,  or {@code Optional.empty} if the table
+   * is not present in the cache or if no expireAfterAccess policy is present in this CachingCatalog.
+   */
+  @VisibleForTesting
+  Optional<Duration> cachedEntryAge(TableIdentifier identifier) {
+    return tableCache.policy()
+        .expireAfterAccess()
+        .flatMap(tableExpiration -> tableExpiration.ageOf(identifier));
+  }
+
+  // Returns the cached Table entry corresponding to the given identifier iff
+  // it's in the cache. Grabs the table in a way that doesn't count as an access
+  // and thus won't affect the cached entry's ttl (if enabled).
+  @VisibleForTesting
+  Optional<Table> tableFromCacheQuietly(TableIdentifier identifier) {
+    // Ensure async cleanup actions have happened.
+    tableCache.cleanUp();
+    return Optional.ofNullable(tableCache.policy().getIfPresentQuietly(identifier));
+  }
+
+  @VisibleForTesting
+  Cache<TableIdentifier, Table> cache() {
+    return tableCache;
+  }
+
+  @VisibleForTesting
+  Optional<Duration> getTimeToTTL(TableIdentifier identifier) {
+    return tableCache
+        .policy()
+        .expireAfterAccess()  // Assumes expireAfterAccess, which is what we set at cache level.
+        .flatMap(tableExpiration -> tableExpiration.ageOf(identifier)) // Get the time the table has been cached.
+        .map(age -> Duration.ofMillis(expirationIntervalMillis).minus(age));
+  }
+
+  private Cache<TableIdentifier, Table> createTableCache(Ticker ticker) {
+    Caffeine<TableIdentifier, Table> cacheBuilder = Caffeine
+        .newBuilder()
+        .softValues()
+        .removalListener(keyLoggingRemovalListener)
+        .writer(new CacheWriter<TableIdentifier, Table>() {
+          @Override
+          // TODO - Consider expiring and syncing any metadata tables that have a different snapshotId
+          //        upon write.
+          public void write(TableIdentifier tableIdentifier, Table table) {
+            LOG.info("Table {} was written to the catalog with snapshotId {}", tableIdentifier,
+                table.currentSnapshot() == null ? null : table.currentSnapshot().snapshotId());
+          }
+
+          @Override
+          public void delete(TableIdentifier tableIdentifier, Table table, RemovalCause cause) {
+            // On expiration, remove any associated metadata tables so that subsequent catalog loads won't
+            // return stale metadata tables w.r.t. the underlying data tables they would return.
+            //
+            // TODO - Should we put metadata tables back into the catalog if their associated table is still
+            //        cached to keep tables and metadata tables on the same snapshot?
+            if (expirationEnabled && !MetadataTableUtils.hasMetadataTableName(tableIdentifier)) {
```

Comment:
The `hasMetadataTableName` check is duplicated in `onTableExpiration`. I think I prefer it there instead of here. You may as well move the `expirationEnabled` there as well. Or maybe this should check the cause and only call `onTableExpiration` if that is why the record is being removed?

Reply:
Good catch. Will be much simpler if it's all in the `onTableExpiration` function.

Reply:
What do you think about checking the cause and using that instead of `expirationEnabled`?

Reply:
I was thinking about that too. Checking expirationEnabled alone isn't necessarily sufficient depending on the behavior we want.

The causes are as follows for those who don't know:

- `EXPLICIT` - Things like `invalidate`, `remove`, etc. 
- `REPLACED` - New entry overwriting an old one - Doesn't seem possible looking at the code.
- `COLLECTED` - for us, this would occur due to configuring `softValues`.
- `EXPIRED` - the reason that would come from `expireAfterAccess etc - what `expirationEnabled` really means.
- `SIZE` - Not relevant to us now as we don't limit cache size.

How we react to each depends on how much we want to try to keep metadata tables in sync with their main table (which currently it does do when `drop` etc are called).

My thoughts on the ones I believe will come up:
- `EXPLICIT`: I think it's possible to ignore since we call `invalidateAll` on metadata tables already on drop etc.
- `COLLECTED`: Whether or not we expire metadata tables when origin table is GC'd I'm very open to ideas on
- `EXPIRED`: Yes, we should all the function.

I can change it to be reacting in a switch statement to the possible enums while we decide on this?

Reply:
Looks like `EXPIRED` allows us to avoid checking `expirationEnabled` so let's plan to do that.

For the others, I think that we could use this as a way to invalidate the cached entries that were dependent on the table. But, that should already be done as you noted for `EXPLICIT`. It is probably a good idea to eventually move this over to happen through the same code path here eventually, but it should be okay to leave it as-is for now.

Reply:
I updated this to use `EXPIRED` and left a note about considering having other expirations happen there. It does admittedly happen in a background thread, but left as a follow up task.